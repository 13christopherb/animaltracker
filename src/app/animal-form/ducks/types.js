//@flow

import {actionTypes} from "./action-types";

export type Animal = {
    id: string,
    name: string,
    weight: string,
    location: string,
    isGettingTubed: boolean,
    isGettingControlledMeds: boolean
}

export type Field = {
    name: string,
    value: string | boolean
}

export type State = {
    +animals: Array<{
        +id: string,
        +name: string,
        +weight: string,
        +location: string,
        +isGettingTubed: boolean,
        +isGettingControlledMeds: boolean
    }>,
    +fields: {
        +name: string,
        +weight: string,
        +location: string,
        +isGettingTubed: boolean,
        +isGettingControlledMeds: boolean
    }
};

export type ChangeFieldAction = {
    type: 'CHANGE_FIELD_VALUE',
    field: {name: string, value: string | boolean}
}

export type AddAnimalSuccessAction = {
    type: 'ADD_ANIMAL_SUCCESS',
    animal: {
        id: string,
        name: string,
        weight: string,
        location: string,
        isGettingTubed: boolean,
        isGettingControlledMeds: boolean}
}

export type AddAnimalFailureAction = {
    type: 'ADD_ANIMAL_FAILURE',
}

export type AddAnimalRequestAction = {
    type: 'ADD_ANIMAL_REQUEST',
    animal: {
        id: string,
        name: string,
        weight: string,
        location: string,
        isGettingTubed: boolean,
        isGettingControlledMeds: boolean}
}

export type Action = ChangeFieldAction | AddAnimalSuccessAction | AddAnimalFailureAction | AddAnimalRequestAction

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;
export type GetState = () => State;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type PromiseAction = Promise<Action>;