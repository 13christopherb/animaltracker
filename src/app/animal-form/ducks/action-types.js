//@flow

export const actionTypes = {
    ADD_ANIMAL_REQUEST: 'ADD_ANIMAL_REQUEST',
    ADD_ANIMAL_SUCCESS: 'ADD_ANIMAL_SUCCESS',
    ADD_ANIMAL_FAILURE: 'ADD_ANIMAL_FAILURE',
    CHANGE_FIELD_VALUE: 'CHANGE_FIELD_VALUE',
};