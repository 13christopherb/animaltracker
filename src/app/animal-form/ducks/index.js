/** @module animal-form */
import animalFormReducer from './reducers';
export {actions as animalFormActions} from './actions';
export {types as animalFormTypes} from './types';
export default animalFormReducer;