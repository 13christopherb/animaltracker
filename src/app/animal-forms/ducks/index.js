/** @module animal-form */
export {types as animalFormTypes} from './types';