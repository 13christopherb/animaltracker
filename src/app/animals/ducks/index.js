import animalsReducer from './reducers';
export {actions as animalsActions} from './actions';
export {actionTypes as animalsTypes} from './action-types';
export default animalsReducer;