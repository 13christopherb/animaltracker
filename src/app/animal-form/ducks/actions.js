//@flow
import * as animalService from "../../animals/ducks/operations";
import type {Animal, Field, ThunkAction, ChangeFieldAction,
    AddAnimalRequestAction, AddAnimalSuccessAction, AddAnimalFailureAction} from './types';
import {actionTypes} from "./action-types";

export const actions = {
    addAnimal,
    changeFieldValue
};

function addAnimal(animal:Animal): ThunkAction {
    return dispatch => {
        dispatch(request(animal));
        return animalService.addAnimal(animal)
            .then(
                res => {
                    dispatch(success(res['animal']))
                },
                error => {
                    dispatch(failure(error))
                }
            );
    };

    function request(animal): AddAnimalRequestAction {
        return {type: actionTypes.ADD_ANIMAL_REQUEST, animal}
    }

    function success(animal): AddAnimalSuccessAction {
        return {type: actionTypes.ADD_ANIMAL_SUCCESS, animal}
    }

    function failure(error): AddAnimalFailureAction {
        return {type: actionTypes.ADD_ANIMAL_FAILURE, error}
    }
}

function changeFieldValue(field:Field): ChangeFieldAction {
    return { type: actionTypes.CHANGE_FIELD_VALUE, field}
}