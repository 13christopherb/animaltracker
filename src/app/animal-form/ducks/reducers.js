//@flow

import {actionTypes} from "./action-types";
import type {State, Action} from './types';

let initialState: State = {
    animals: [],
    fields: {
        name: '',
        weight: '',
        location: '',
        isGettingTubed: false,
        isGettingControlledMeds: false
    }
};

export default function animalFormReducer(state:State=initialState, action:Action): State {
    switch (action.type) {
        case actionTypes.CHANGE_FIELD_VALUE:
            return {
                ...state,
                fields: {
                    ...state.fields,
                    [action.field.name]: action.field.value
                }
            };
        default:
            return state
    }
}