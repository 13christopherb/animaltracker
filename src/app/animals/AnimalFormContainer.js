import React, {Component} from 'react';
import {connect} from 'react-redux';
import moment from 'moment';
import {actions} from './ducks/Actions';
import {AnimalForm} from "./AnimalForm";

class AnimalFormContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            value: '',
            isGettingTubed: false,
            isGettingControlledMeds: false,
            name: '',
            weight: '',
            species: '',
            location: '',
            formErrors: {name: '', weight: '', species: '', location: ''},
            isNameValid: false,
            isWeightValid: false,
            isSpeciesValid: false,
            isLocationValid: false,
            isFormValid: false
        };
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    componentDidMount() {
        //this.props.dispatch(actions.fetchAnimals());
    }

    /**
     * Saves the input values to the state to submit later
     * when button is pressed
     * @param e Input change event
     */
    handleInputChange(e) {
        const target = e.target;
        const value = target.type === 'checkbox' ? target.checked : target.value;
        const name = target.name;
        this.setState({
            [name]: value
        });
    }

    /**
     * post the input values saved in the state
     * to the server.
     * @param e On click event
     */
    handleSubmit(e) {
        e.preventDefault();
        const animal = {
            timestamp: moment().format(),
            location: this.state.location,
            name: this.state.name,
            species: this.state.species,
            weight: this.state.weight,
            isGettingControlledMeds: this.state.isGettingControlledMeds,
            isGettingTubed: this.state.isGettingTubed
        };
        this.props.dispatch(actions.addAnimal(animal));
    }

    render() {
        return (
            <div>
                <AnimalForm
                    handleInputChange={this.handleInputChange}
                    handleSubmit={this.handleSubmit}
                    location={this.state.location}
                    name={this.state.name}
                    species={this.state.species}
                    weight={this.state.weight}
                    isGettingControlledMeds={this.state.isGettingControlledMeds}
                    isGettingTubed={this.state.isGettingTubed}
                    toggleAddAnimal={this.props.toggleAddAnimal}
                />
            </div>
        );
    }
}

function mapStateToProps({animals}, ownProps) {
    return {
        animals: animals
    }
}

export default connect(
    mapStateToProps,
)(AnimalFormContainer);