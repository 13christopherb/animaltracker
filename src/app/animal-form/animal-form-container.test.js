import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk'
import AnimalFormContainer from "./animal-form-container";
import {animalsTypes} from '../animals/ducks/'

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const initialState = {
    animals: [],
    animalForm: {fields: {}}
};

const store = mockStore(initialState);


import renderer from 'react-test-renderer'
import {Provider} from 'react-redux'
import {reduxForm} from 'redux-form'
import {mount} from "enzyme/build";
import {types} from "../transports/ducks/types";

jest.mock('react-dom');

const spy = jest.fn();


const formFieldValues = {name: 'Edginald'};

describe('<AnimalFormContainer>', () => {
    test('AnimalFormContainer renders correctly', () => {
        const Decorated = reduxForm({
            form: 'animal-form', onSubmit: {spy}
        })(AnimalFormContainer);
        const store = mockStore(initialState);
        const tree = renderer.create(
            <Provider store={store}>
                <Decorated
                    {...formFieldValues}
                />
            </Provider>
        ).toJSON();
        expect(tree).toMatchSnapshot()
    });
});
