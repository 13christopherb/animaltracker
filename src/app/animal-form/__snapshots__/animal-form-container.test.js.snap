// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimalFormContainer> AnimalFormContainer renders correctly 1`] = `
<div>
  <div>
    <div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-row"
        >
          <div
            className="col col-md-2"
          >
            <div
              className="form-group"
            >
              <label>
                Location
              </label>
              <select
                className="custom-select "
                id="location"
                name="location"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                value=""
              >
                <option
                  disabled={true}
                  value=""
                />
                <option
                  label="NRO"
                  value="NRO"
                >
                  NRO
                </option>
                <option
                  label="MBO"
                  value="MBO"
                >
                  MBO
                </option>
                <option
                  label="SLO"
                  value="SLO"
                >
                  SLO
                </option>
              </select>
            </div>
          </div>
          <div
            className="col col-md-2"
          >
            <div
              className="form-group"
            >
              <label>
                Species
              </label>
              <select
                className="custom-select "
                id="species"
                name="species"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                value=""
              >
                <option
                  disabled={true}
                  value=""
                />
                <option
                  label="CSL"
                  value="CSL"
                >
                  CSL
                </option>
                <option
                  label="ES"
                  value="ES"
                >
                  ES
                </option>
                <option
                  label="HS"
                  value="HS"
                >
                  HS
                </option>
                <option
                  label="NFS"
                  value="NFS"
                >
                  NFS
                </option>
                <option
                  label="GFS"
                  value="GFS"
                >
                  GFS
                </option>
                <option
                  label="SSO"
                  value="SSO"
                >
                  SSO
                </option>
                <option
                  label="CET"
                  value="CET"
                >
                  CET
                </option>
              </select>
            </div>
          </div>
          <div
            className="col col-md-3"
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
              >
                Name
              </label>
              <input
                className="form-control "
                id="name"
                name="name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="col col-md-3"
          >
            <div
              className="form-group"
            >
              <label
                className="form-label"
              >
                Weight
              </label>
              <input
                className="form-control "
                id="weight"
                name="weight"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="form-row"
        >
          <div
            className="col col-md-6"
          >
            <label>
              Controlled meds:
              <input
                checked={false}
                name="isGettingControlledMeds"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                type="checkbox"
                value=""
              />
            </label>
            <label>
              Tube feeding:
              <input
                checked={false}
                name="isGettingTubed"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                type="checkbox"
                value=""
              />
            </label>
            <button
              className="btn btn-primary"
              disabled={true}
              type="submit"
            >
              Submit
            </button>
            <button
              className="btn btn-danger"
              type="button"
            >
              Cancel
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
